helpers/compile/plugins/tscPlugin.ts

```diff
@@ -1,23 +1,101 @@
+        '@prisma/internals',
+        '@prisma/engine-core',
+        '@prisma/generator-helper',
+        '@prisma/debug',
+      ],
+      compiler: {
+        tsconfigFilePath: path.join(process.cwd(), 'tsconfig.build.json'),
+        overrideTsconfig: {
+          compilerOptions: {
+            paths: {}, // bug with api extract + paths
+          },
+        },
+      },
+      dtsRollup: {
+        enabled: true,
+        untrimmedFilePath: path.join(process.cwd(), `${outfile}.d.ts`),
+      },
+      tsdocMetadata: {
+        enabled: false,
+      },
+    },
+    packageJsonFullPath: path.join(process.cwd(), 'package.json'),
+    configObjectFullPath: undefined,
+  })
+
+  // here we trigger the "command line" interface equivalent
+  const extractorResult = Extractor.invoke(extractorConfig, {
+    showVerboseMessages: true,
+    localBuild: true,
+  })
+
+  // we exit the process immediately if there were errors
+  if (extractorResult.succeeded === false) {
+    console.error(`API Extractor completed with errors`)
+    process.exit(1)
+  }
+}
+
+/**
+ * Triggers the TypeScript compiler and the type bundler.
+ */
+export const tscPlugin: (emitTypes?: boolean) => esbuild.Plugin = (emitTypes?: boolean) => ({
   name: 'tscPlugin',
   setup(build) {
     const options = build.initialOptions
 
-    if (process.env.DEV === 'true') return
+    if (emitTypes === false) return // build has opted out of emitting types
 
     build.onStart(async () => {
-      // --paths null basically prevents typescript from using paths from the
-      // tsconfig.json that is passed from the esbuild config. We need to do
-      // this because TS would include types from the paths into this build.
-      // But our paths, in our specific case only represent separate packages.
-      await run(`tsc --project ${options.tsconfig} --paths null`)
+      // we only call tsc if not in watch mode or in dev mode (they skip types)
+      if (options.incremental !== true && process.env.DEV !== 'true') {
+        // --paths null basically prevents typescript from using paths from the
+        // tsconfig.json that is passed from the esbuild config. We need to do
+        // this because TS would include types from the paths into this build.
+        // but our paths, in our specific case only represent separate packages.
+        await run(`tsc --project ${options.tsconfig} --paths null`)
+      }
+
+      // we bundle types if we also bundle the entry point and types exists
+      if (options.bundle === true && options.entryPoints?.[0].endsWith('.ts')) {
+        const tsconfig = require(`${process.cwd()}/${options.tsconfig}`) // tsconfig
+        const typeOutDir = tsconfig?.compilerOptions?.outDir ?? '.' // type out dir
+        const entryPoint = options.entryPoints?.[0].slice(0, -3) // drop ".ts"
+        const typeOutPath = `${typeOutDir}/${entryPoint.slice(4)}` // drop "src/"
+        const bundlePath = options.outfile!.slice(0, -3) // drop ".ts"
```

Comment:
```suggestion
        const typeOutDir = tsconfig?.compilerOptions?.outDir ?? '.' // type out dir
        const entryPoint = options.entryPoints?.[0].replace(/\.ts$/, '')
        const typeOutPath = `${typeOutDir}/${entryPoint.replace(/^src\//, '')}` // drop "src/"
        const bundlePath = options.outfile!.replace(/\.ts$/, '')
```

Do not insist on this change, but in my opinion that makes intention a little bit clearer and protects us in case any of the parts we replace change. Feel free to ignore if you disagree.

Reply:
I think that's better indeed, and removes the need for comments.